File: raspPi/src/nodes/NodeFactory.cpp
#include "NodeFactory.hpp"
#include "MirrorNode.hpp"
#include <typeinfo>
#include <typeindex>
#include <array>
#include <tuple>

using namespace std;

static const array< tuple<type_index, std::string>, 11> typeNames =
{
    std::make_tuple(std::type_index(typeid(BaseNode)),       "node"),
    std::make_tuple(std::type_index(typeid(MirrorNode)),     "mirror"),
    std::make_tuple(std::type_index(typeid(BoolNode)),       "bool"),
    std::make_tuple(std::type_index(typeid(UInt16Node)),     "uint16"),
    std::make_tuple(std::type_index(typeid(UInt32Node)),     "uint32"),
    std::make_tuple(std::type_index(typeid(Int32Node)),      "int32"),
    std::make_tuple(std::type_index(typeid(Int64Node)),      "int64"),
    std::make_tuple(std::type_index(typeid(FloatNode)),      "float"),
    std::make_tuple(std::type_index(typeid(DoubleNode)),     "double"),
    std::make_tuple(std::type_index(typeid(Pos3D_32f_Node)), "pos3D_32f"),
    std::make_tuple(std::type_index(typeid(StringNode)),     "string")
};
std::string unknownNode = "Unknown";

std::unique_ptr<BaseNode> NodeFactory::CreateNode(const std::string& nodeType, const std::string& nodeName)
{
    for(const auto& type : typeNames)
    {
        if (get<1>(type).compare(nodeType) != 0)
        {
            continue;
        }
        const type_index& typeIdx = get<0>(type);
#define CREATE_NODE(TNodeType) if (typeIdx == typeid(TNodeType)) { return make_unique<TNodeType>(nodeName); }

        CREATE_NODE(BaseNode);
        CREATE_NODE(MirrorNode);
        CREATE_NODE(BoolNode);
        CREATE_NODE(UInt16Node);
        CREATE_NODE(UInt32Node);
        CREATE_NODE(Int32Node);
        CREATE_NODE(Int64Node);
        CREATE_NODE(FloatNode);
        CREATE_NODE(DoubleNode);
        CREATE_NODE(Pos3D_32f_Node);
        CREATE_NODE(StringNode);
#undef CREATE_NODE
    }
    return nullptr;
    // Error unknown node type.
}

const std::string& NodeFactory::NodeTypeName(const BaseNode* pNode)
{
    for (const auto& type : typeNames)
    {
        if (get<0>(type) == type_index(typeid(*pNode)) )
        {
            return get<1>(type);
        }
    }
    // Error unknown node type.
    return unknownNode;
}

File: ROBINsim/Assets/Scripts/nodes/NodeFactory.cs
using System;

namespace ROBINinspect
{
    public enum NodeTypes {
        nodeNode,
        mirrorNode,
        boolNode,
        uint16Node,
        uint32Node,
        int32Node,
        int64Node,
        floatNode,
        doubleNode,
        pos3D_32fNode,
        stringNode
    }
    class NodeFactory
    {
        private static String[] nodeNames;
        private static String[] NodeNames
        {
            get
            {
                if (nodeNames == null)
                {
                    Array enumArr = Enum.GetValues(typeof(NodeTypes));
                    nodeNames = new String[enumArr.Length];
                    foreach (NodeTypes eNodeType in enumArr)
                    {
                        String tmp = eNodeType.ToString();
                        nodeNames[(int)eNodeType] = tmp.Replace("Node", "");
                    }
                }
                return nodeNames;
            }
        }
        public static BaseNode CreateNode(String nodeType, String nodeName)
        {
            BaseNode newNode = null;
            NodeTypes eNodeType = NodeTypes.nodeNode;
            if(GetNodeType(nodeType, ref eNodeType))
            {
                switch (eNodeType)
                {
                    case NodeTypes.nodeNode:
                        newNode = new BaseNode(nodeName);
                        break;
                    case NodeTypes.mirrorNode:
                        newNode = new MirrorNode(nodeName);
                        break;                        
                    case NodeTypes.boolNode:
                        newNode = new BoolNode(nodeName);
                        break;
                    case NodeTypes.uint16Node:
                        newNode = new UInt16Node(nodeName);
                        break;
                    case NodeTypes.uint32Node:
                        newNode = new UInt32Node(nodeName);
                        break;
                    case NodeTypes.int32Node:
                        newNode = new Int32Node(nodeName);
                        break;
                    case NodeTypes.int64Node:
                        newNode = new Int64Node(nodeName);
                        break;
                    case NodeTypes.floatNode:
                        newNode = new FloatNode(nodeName);
                        break;
                    case NodeTypes.doubleNode:
                        newNode = new DoubleNode(nodeName);
                        break;
                    case NodeTypes.pos3D_32fNode:
                        newNode = new Pos3D_32f_Node(nodeName);
                        break;
                    case NodeTypes.stringNode:
                        newNode = new StringNode(nodeName);
                        break;
                    default:
                        break;
                }
            }        
            return newNode;
        }

        static public bool GetNodeType(String strNodeType, ref NodeTypes type)
        {
            String[] names = NodeNames;
            int i = 0;
            foreach (String str in names)
            {
                if (str.Equals(strNodeType))
                {
                    type = (NodeTypes)i;
                    return true;
                }
                i++;
            }
            return false;
        }

        static public String TypeAsString(NodeTypes type)
        {
            if ((int)type >= 0 && (int)type < NodeNames.Length)
                return NodeNames[(int)type];
            return "Unknown";
        }

        static public NodeTypes NodeType(BaseNode node)
        {

            if (node.GetType() == typeof(BaseNode)) return NodeTypes.nodeNode;
            if (node.GetType() == typeof(BoolNode)) return NodeTypes.boolNode;
            if (node.GetType() == typeof(UInt16Node)) return NodeTypes.uint16Node;
            if (node.GetType() == typeof(UInt32Node)) return NodeTypes.uint32Node;
            if (node.GetType() == typeof(Int32Node)) return NodeTypes.int32Node;
            if (node.GetType() == typeof(Int64Node)) return NodeTypes.int64Node;
            if (node.GetType() == typeof(FloatNode)) return NodeTypes.floatNode;
            if (node.GetType() == typeof(DoubleNode)) return NodeTypes.doubleNode;
            if (node.GetType() == typeof(Pos3D_32f_Node)) return NodeTypes.pos3D_32fNode;
            if (node.GetType() == typeof(StringNode)) return NodeTypes.stringNode;
            if (node.GetType() == typeof(MirrorNode)) return NodeTypes.mirrorNode;
            throw new NotSupportedException("You need to add support for node" + node.GetType().ToString());
        }
    }
}
